File: algorithms/2_medium/M/03443_maximum_manhattan_distance_after_k_changes/solution.c
static inline int max_int(const int a, const int b)
{
    return a > b ? a : b;
}

struct data {
    int curr;
    char a;
    char b;
    int k;
};

static inline struct data data_init(const char a, const char b, const int k)
{
    return (struct data){ .curr = 0, .a = a, .b = b, .k = k };
}

__attribute__((no_sanitize("all")))
static inline void data_add(struct data* obj, const char ch)
{
    if (ch == obj->a || ch == obj->b) {
        obj->curr += 1;
    } else if (obj->k > 0) {
        obj->k -= 1;
        obj->curr += 1;
    } else {
        obj->curr -= 1;
    }
}

__attribute__((no_sanitize("all")))
int maxDistance(const char* s, const int k)
{
    if (s == NULL) { return 0; }
    int max = 0;
    struct data ne = data_init('N', 'E', k);
    struct data nw = data_init('N', 'W', k);
    struct data se = data_init('S', 'E', k);
    struct data sw = data_init('S', 'W', k);

    for (const char* it = s; *it != '\0'; it += 1) {
        assert(it - s <= 100000);

        data_add(&ne, *it);
        data_add(&nw, *it);
        data_add(&se, *it);
        data_add(&sw, *it);

        max = max_int(max, ne.curr);
        max = max_int(max, nw.curr);
        max = max_int(max, se.curr);
        max = max_int(max, sw.curr);
    }

    return max;
}

File: algorithms/2_medium/M/03443_maximum_manhattan_distance_after_k_changes/solution.rs
struct Data {
    curr: i32,
    a: u8,
    b: u8,
    k: i32,
}

impl Data {
    #[inline(always)]
    fn new(a: u8, b:u8, k: i32) -> Self {
        Self { curr: 0, a: a, b: b, k: k }
    }

    #[inline(always)]
    fn add(&mut self, byte: u8) {
        if byte == self.a || byte == self.b {
            self.curr += 1;
        } else if self.k > 0 {
            self.k -= 1;
            self.curr += 1;
        } else {
            self.curr -= 1;
        }
    }
}

impl Solution {
    pub fn max_distance(s: String, k: i32) -> i32 {
        assert!(s.len() <= 100_000);

        let mut max = 0_i32;
        let mut ne = Data::new(b'N', b'E', k);
        let mut nw = Data::new(b'N', b'W', k);
        let mut se = Data::new(b'S', b'E', k);
        let mut sw = Data::new(b'S', b'W', k);

        for &byte in s.as_bytes() {
            ne.add(byte);
            nw.add(byte);
            se.add(byte);
            sw.add(byte);

            max = max.max(ne.curr).max(nw.curr).max(sw.curr).max(se.curr);
        }

        max
    }
}
